### PR TITLE
[Chore-3] Setup base HTTP service

### LINKED ISSUE #30: [Chore-3] Setup base HTTP service
- [x] base HTTP service
- [x] handle error response

### PR DESCRIPTION
**Pull request by carryall**
## What happened

Solved #30 

- use [custom webpack builder](https://github.com/just-jeb/angular-builders/tree/master/packages/custom-webpack) for Angular to add custom configuration so that `process.env` can be used to get environment variables
- add `BaseService` to handle API requests
- use [ts-jsonapi](https://github.com/mohuk/ts-jsonapi) to deserialize Json API response

Noted:
- base API request with authenticated header would be implemented in another PR for #32 

## Insight

N/A

## Proof Of Work

- all specs passed
- tested locally that environment variable works by log the ENV values on any component
```ts
import { environment } from 'src/environments/environment';
//...
console.log(environment.apiBaseUrl);
console.log(environment.apiClientID);
console.log(environment.apiClientSecret);
console.log(environment.apiVersion);
```
Screen Shot 2563-11-27 at 14 48 54

### CODE REVIEW

File: src/app/shared/services/base.service.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import { environment } from '../../../environments/environment';
+import { HttpClient, HttpErrorResponse } from '@angular/common/http';
+import { catchError, retry } from 'rxjs/operators';
+import { map } from 'rxjs/operators';
+import { Deserializer } from 'ts-jsonapi';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class BaseService {
```
Comment (junan) [inline src/app/shared/services/base.service.ts]: @carryall How about to make it `abstract` class as this class will be used for inheritance and then no one cant cannot directly create an instance of an abstract class?

File: package.json [hunk 2]
```diff
@@ -30,18 +30,22 @@
...
     "webdriver-manager": "12.1.7",
     "zone.js": "0.10.3"
   },
   "devDependencies": {
+    "@angular-builders/custom-webpack": "10.0.1",
     "@angular-devkit/build-angular": "0.1100.0",
     "@angular/cli": "11.0.0",
     "@angular/compiler-cli": "11.0.0",
     "@types/jasmine": "3.6.1",
+    "@types/lodash": "^4.14.165",
```
Comment (junan) [inline package.json]: @carryall Could we use specific version just like others?

File: src/app/shared/services/base.service.ts (hunk 1 above)
Comment (junan) [inline src/app/shared/services/base.service.ts]: @carryall Could we add extra space before this line?

Comment (carryall) [reply src/app/shared/services/base.service.ts]: ✅  Fixed on [e8d6d6e](https://github.com/carryall/survey-challenge/pull/31/commits/e8d6d6eec9da251ee7dead5965815559c3d97062)

File: package.json (hunk 2 above)
Comment (carryall) [reply package.json]: ✅  Fixed on [e8d6d6e](https://github.com/carryall/survey-challenge/pull/31/commits/e8d6d6eec9da251ee7dead5965815559c3d97062)

File: src/app/shared/services/base.service.ts (hunk 1 above)
Comment (carryall) [reply src/app/shared/services/base.service.ts]: ✅  Fixed on [e8d6d6e](https://github.com/carryall/survey-challenge/pull/31/commits/e8d6d6eec9da251ee7dead5965815559c3d97062)

**Review by junan — APPROVED** [review]

File: .env.sample [hunk 3]
```diff
@@ -0,0 +1,4 @@
+API_BASE_URL="https://survey-api-staging.nimblehq.co"
```
Comment (olivierobert) [inline .env.sample]: Why not using the production environment instead? Staging can break if we make changes.

Comment (olivierobert) [inline .env.sample]: Quotes can be omitted in env files:

```
API_BASE_URL=https://survey-api-staging.nimblehq.co
```

File: src/app/shared/services/base.service.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  post(endpoint: string, body: {}): Observable<any> {
+    const apiUrl = this.apiUrlFor(endpoint);
+
+    return this.http.post(apiUrl, body).pipe(
+      retry(1),
+      catchError(this.handleError),
+      map(response => this.deserialize(response)));
+  }
+
+  protected handleError(error: any): Observable<any> {
```
Comment (olivierobert) [inline src/app/shared/services/base.service.ts]: Why not `private`? Would there be a case when this method would be overridden in a child class?

File: src/app/shared/services/base.service.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+  deserializer: Deserializer;
+
+  constructor(private http: HttpClient) {
+    this.deserializer = new Deserializer({
+      keyForAttribute: 'camelCase'
+    });
+  }
+
+  post(endpoint: string, body: {}): Observable<any> {
```
Comment (olivierobert) [inline src/app/shared/services/base.service.ts]: Why is `Observable<any>` needed? I want to learn 😁

File: src/environments/environment.prod.ts [hunk 6]
```diff
@@ -1,3 +1,7 @@
 export const environment = {
-  production: true
+  production: true,
```
Comment (olivierobert) [inline src/environments/environment.prod.ts]: How about using a predicate-like method naming: 

```suggestion
  isProduction: true,
```

File: tsconfig.app.json [hunk 7]
```diff
@@ -3,7 +3,7 @@
   "extends": "./tsconfig.json",
   "compilerOptions": {
     "outDir": "./out-tsc/app",
-    "types": []
+    "types": ["node"]
```
Comment (olivierobert) [inline tsconfig.app.json:6]: Why is this needed?

File: src/app/shared/services/base.service.spec.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        const deserializedResponse = {
+          id: 10,
+          accessToken: 'lbxD2K2BjbYtNzz8xjvh2FvSKx838KBCf79q773kq2c',
+          tokenType: 'Bearer',
+          expiresIn: 7200,
+          refreshToken: '3zJz2oW0njxlj_I3ghyUBF7ZfdQKYXd2n0ODlMkAjHc',
+          createdAt: 1597169495
+        };
+
+        const data = {
```
Comment (olivierobert) [inline src/app/shared/services/base.service.spec.ts]: I was a bit confused by the naming `data` as I thought initially it referred to the data attributethe JSON response. So I would recommend to make this variable name more specific e.g. `postRequestPayload`.

File: src/app/shared/services/base.service.ts [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    const apiUrl = this.apiUrlFor(endpoint);
+
+    return this.http.post(apiUrl, body).pipe(
```
Comment (olivierobert) [inline src/app/shared/services/base.service.ts]: Referring to my above comment, what do you think of renaming `body` to `payload`? It's slightly more descriptive and less generic. It's a minor naming improvement but combined with the tests, I find it clearer.

File: src/app/shared/services/base.service.spec.ts [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    service = TestBed.inject(BaseService);
+    httpMock = TestBed.inject(HttpTestingController);
+  });
+
+  it('creates the service', () => {
+    expect(service).toBeTruthy();
+  });
+
+  describe('#post', () => {
+    describe('Given a valid data', () => {
```
Comment (olivierobert) [inline src/app/shared/services/base.service.spec.ts]: `data` is always too generic and not precise. So I usually recommend o stay clear from it unless there are no better options. Here I find it confusing as `data` could refer to both the JSON API attribute and request body. How about to clarify with:

```suggestion
    describe('Given a valid request payload', () => {
```

File: src/app/shared/services/base.service.spec.ts [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+          client_id: environment.apiClientID,
+          client_secret: environment.apiClientSecret
+        };
+
+        service.post('oauth/token', data).subscribe(response => {
+          expect(response).toEqual(deserializedResponse);
+        });
+
+        const request = httpMock.expectOne(`${environment.apiBaseUrl}/api/${environment.apiVersion}/oauth/token`);
+        expect(request.request.method).toBe('POST');
```
Comment (olivierobert) [inline src/app/shared/services/base.service.spec.ts:65]: Should we also make sure that there are no pending unhandled requests?

```javascript
httpMock.verify()
```

Not sure if it's a best practice in Angular JS but when using Mocks, it's usually better to make sure nothing is left unhandled.

File: src/app/shared/services/base.service.spec.ts [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+        const data = {
+          grant_type: 'password',
+          email: 'invalid_email@nimblehq.co',
+          password: '12345678',
+          client_id: environment.apiClientID,
+          client_secret: environment.apiClientSecret
+        };
+
+        service.post('oauth/token', data).subscribe(
+          _ => fail('Should have failed with 400 error'),
```
Comment (olivierobert) [inline src/app/shared/services/base.service.spec.ts:91]: Will that be the error message if this expectation is not met?

File: src/app/shared/services/base.service.spec.ts [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+        });
+
+        const request = httpMock.expectOne(`${environment.apiBaseUrl}/api/${environment.apiVersion}/oauth/token`);
+        expect(request.request.method).toBe('POST');
+        request.flush(mockResponse);
+      });
+    });
+
+    describe('Given an INVALID data', () => {
+      it('throws error', () => {
```
Comment (olivierobert) [inline src/app/shared/services/base.service.spec.ts]: English:

```suggestion
      it('throws an error', () => {
```

Comment (olivierobert) [inline src/app/shared/services/base.service.spec.ts]: English:

```suggestion
      it('throws an error', () => {
```

**Review by olivierobert — COMMENTED** [review]
English syntax improvements in the PR description:

> use customer webpack builder for Angular to add custom configuration so we can use process.env to get environment variables

But you mean:

> use custom webpack builder for Angular to add custom configuration so that process.env can be used to get environment variables

File: src/app/shared/services/base.service.spec.ts (hunk 12 above)
Comment (carryall) [reply src/app/shared/services/base.service.spec.ts:91]: @olivierobert  Yes the error message on testing would look like this
Screen Shot 2563-12-10 at 12 06 57

File: src/app/shared/services/base.service.ts (hunk 5 above)
Comment (carryall) [reply src/app/shared/services/base.service.ts]: @olivierobert So when we use the services we can `.subscribe` for response and error later 🤔

File: tsconfig.app.json (hunk 7 above)
Comment (carryall) [reply tsconfig.app.json:6]: @olivierobert To be able to use `process.env` on environment files because `process` is from node

**Review by olivierobert — APPROVED** [review]
